moshi/src/main/java/com/squareup/moshi/LinkedHashTreeMap.kt

```diff
@@ -0,0 +1,775 @@
+package com.squareup.moshi
+
+import com.squareup.moshi.LinkedHashTreeMap.Node
+import com.squareup.moshi.internal.knownNotNull
+import java.io.Serializable
+import kotlin.math.max
+
+@Suppress("UNCHECKED_CAST")
+private val NATURAL_ORDER = Comparator<Any> { o1, o2 -> (o1 as Comparable<Any>).compareTo(o2) }
+
+/**
+ * A map of comparable keys to values. Unlike TreeMap, this class uses insertion order for
+ * iteration order. Comparison order is only used as an optimization for efficient insertion and
+ * removal.
+ *
+ * This implementation was derived from Android 4.1's TreeMap and LinkedHashMap classes.
+ */
+internal class LinkedHashTreeMap<K, V>
+/**
+ * Create a tree map ordered by [comparator]. This map's keys may only be null if [comparator] permits.
+ *
+ * @param comparator the comparator to order elements with, or null to use the natural
+ *     ordering.
```

Comment:
```suggestion
 * @param comparator the comparator to order elements with, or null to use the natural ordering.
```

Reply:
Hi Gooler, this is a needless formatting suggestion as it is correct as is. Since this is a complex PR I'd like to ask that you avoid trying to contribute further changes like this as they just add noise that distract from the more important parts of the review.